File: handlers/bookcover.ts
import { Request, Response, NextFunction } from 'express';
import { getLinkGoogle, getLinkGoodreads } from '../helpers/bookcover';
import { BOOKCOVER_NOT_FOUND, INVALID_ISBN } from '../helpers/messages';
import Axios from 'axios';
import Url from 'url';
import HttpException from '../exceptions/HttpException';

type BookcoverResponse = {
  status: string,
  url: string
  bookTitle?: string,
  authorName?: string,
  isbn?: string,
}

export const getBookcoverUrl = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const query = new URLSearchParams(Url.parse(req.url).query);
    const bookTitle = query.get('book_title');
    const authorName = query.get('author_name');
    const googleQuery = `${bookTitle} ${authorName} site:goodreads.com/book/show`;
    const googleResponse = await Axios.get(`https://www.google.com/search?q=${googleQuery}&sourceid=chrome&ie=UTF-8`);

    const goodreadsLink = getLinkGoogle(googleResponse.data);
    if (!goodreadsLink) {
      throw new HttpException(404, BOOKCOVER_NOT_FOUND);
    }

    const goodreadsResponse = await Axios.get(goodreadsLink);
    return res.json({ url: getLinkGoodreads(goodreadsResponse.data) });
  } catch (error) {
    next(error);
  }
}

export const getBookcoverFromISBN = async (req: Request, res: Response, next: NextFunction) => {
  try {
    const isbn = req.params.id.replace(/-/g, '');
    if (isbn.length !== 13) {
      throw new HttpException(400, INVALID_ISBN);
    }

    const response = await Axios.get(
      `https://www.googleapis.com/books/v1/volumes?q=isbn:${isbn}&key=${process.env.GOOGLE_BOOKS_API_KEY}`
    );
    if (!response.data.totalItems) {
      return res.json([]);
    }

    return res.json({ url: response.data.items[0].volumeInfo.imageLinks.thumbnail });
  } catch (error) {
    next(error);
  }
};

File: handlers/bookcover.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

const GOOGLE_BOOKS_API_KEY = "GOOGLE_BOOKS_API_KEY"
const BOOK_TITLE = "book_title"
const AUTHOR_NAME = "author_name"
const START_PATTERN_IMAGE_SEARCH = "https://images-na.ssl-images-amazon.com/images"
const START_PATTERN_GOODREADS_SEARCH = "https://www.goodreads.com/book/show/"
const END_PATTERN_GOODREADS_SEARCH = "&"
const END_PATTERN_IMAGE_SEARCH = "\""

func BookcoverSearch(w http.ResponseWriter, r *http.Request) {
  bookTitle := strings.ReplaceAll(r.URL.Query().Get(BOOK_TITLE), " ", "+")
  authorName := strings.ReplaceAll(r.URL.Query().Get(AUTHOR_NAME), " ", "+")
  q := bookTitle + "+" + authorName + "site:goodreads.com/book/show"
  query := "https://www.google.com/search?q=" + q + "&sourceid=chrome&ie=UTF-8"

  goodreadUrl := getUrl(getBody(w, query), START_PATTERN_GOODREADS_SEARCH, END_PATTERN_GOODREADS_SEARCH)
  imageUrl := getUrl(getBody(w, goodreadUrl), START_PATTERN_IMAGE_SEARCH, END_PATTERN_IMAGE_SEARCH)

  w.Write(BuildSuccessResponse(w, imageUrl))
}

func BookcoverByIsbn(w http.ResponseWriter, r *http.Request) {
  isbn := strings.ReplaceAll(r.PathValue("isbn"), "-", "")
  if(len(isbn) != 13) {
    w.Write(BuildErrorResponse(w, HttpException{ statusCode: http.StatusBadRequest, message: INVALID_ISBN }))
    return
  }

  query := "https://www.googleapis.com/books/v1/volumes?q=isbn:" + isbn + "&key" + os.Getenv(GOOGLE_BOOKS_API_KEY)
  res := getBody(w, query)
  var googleBook GoogleBook = GoogleBook{} 
  if err := json.Unmarshal(res, &googleBook); err != nil {
    fmt.Println("Error while parsing JSON body")
    return
  }

  if googleBook.TotalItems == 0 {
    w.Write(BuildErrorResponse(w, HttpException{ 
      statusCode: http.StatusBadRequest,
      message: BOOKCOVER_NOT_FOUND,
    }))
    return
  }

  responseData := BuildSuccessResponse(w, googleBook.Items[0].VolumeInfo.ImageLinks.Thumbnail)
  w.Write(responseData)
}

func getBody(w http.ResponseWriter, url string) []byte {
  response, err := http.Get(url)
  if err != nil {
    w.Write(BuildErrorResponse(w, HttpException{
      statusCode: http.StatusBadRequest, message: BOOKCOVER_NOT_FOUND,
    }))
  }

  body, err := io.ReadAll(response.Body)
  if err != nil {
    w.Write(BuildErrorResponse(w, HttpException{
      statusCode: http.StatusBadRequest, message: ERROR_READING_BODY,
    }))
  }

  return body
}

func getUrl(data []byte, startPattern string, endPattern string) string {
  body := string(data)
  init := strings.Index(body, startPattern)
  end := strings.Index(body[init:], endPattern)

  return body[init:init+end]
}
